README.md
##introduction

Server implements communication device<=>server [protocol](https://docs.google.com/spreadsheets/d/15s-2ZbqOQ1bZvAtFFm9sIEuKy3jbJzxdeynp72sjoYU/edit#gid=3) 

device: gps tracker [Autolink II](http://tn-group.net/index.php?route=product/product&path=25_29&product_id=68)

## Usage

```clojure
(use 'gpsservices.core)

(defrecord MyEvents []
  ISockEvents
  (on-error [this session err]
    (println "ISockEvents-error:" (.getMessage err)))
  (on-open [this session]
    (println "ISockEvents-open:"))
  (on-message [this session msg]
    (println "ISockEvents-message:" msg))
  (on-close [this session]
    (println "ISockEvents-close: car-id " @(:car-id session))))
    
(start-alink2-server (->MyEvents) 7779) ;; 7779 - listen port
````


msg data is hashmap:

```clojure
   {:type :package,
    :pack-num 2,
    :unixtime 1406097384,
    :data [{:point {:lat 55.0224, :lon 82.9139},
            :status-bytes nil,
            :speed 0.0,
            :satellite-count 20,
            :altitude 130,
            :course 0}]}
````



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased][unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2016-01-20
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2016-01-20
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[unreleased]: https://github.com/your-name/gpsservices/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/gpsservices/compare/0.1.0...0.1.1


test/gpsservices/core_test.clj
(ns gpsservices.core-test
  (:require [clojure.test :refer :all]
            [gpsservices.core :refer :all]))


;;todo more tests


test/gpsservices/autolink2_test.clj
(ns gpsservices.autolink2-test
  (:require [clojure.test :refer :all]
            [gpsservices.autolink2 :refer [get-type-package]]))

(def ^:private exmpl-header (byte-array [0xff 0x22 0x41 0x04 0x55 0x7f 0xa0 0x15 0x03 0x00]))

(def ^:private exmpl-package (byte-array [0x5b 0x02
                                          ;;type pack
                                          0x01
                                          ;;length data
                                          0x23 0x00
                                          ;;unixtime
                                          0xe8 0x57 0xcf 0x53
                                          ;;data start
                                          0x03 0xf0 0x16 0x5c 0x42 0x04 0xeb 0xd3
                                          0xa5 0x42 0x05 0x00 0x0d 0x55 0x00 0x09
                                          0x00 0xf0 0x44 0x5e 0x08 0x63 0xfa 0x00
                                          0x0e 0xfa 0x46 0x00 0x00 0x00 0xfa 0x1f
                                          0x00 0x00 0x00 ;;data end
                                          0x7f           ;;control sum
                                          0x5d]))

(def ^:private exmpl-package-invalid (byte-array [0x5b 0x02
                                                  0x01
                                                  0x23 0x00
                                                  0xe8 0x57 0xcf 0x53
                                                  0x03 0xf0 0x16 0x5c
                                                  0x42 0x04 0xeb 0xd3 0xa5 0x42 0x05 0x00 0x0d 0x55 0x00 0x09 0x00
                                                  0xf0 0x44 0x5e 0x08 0x63 0xfa 0x00 0x0e 0xfa 0x46 0x00 0x00 0x00]))


(deftest detect-type-test-1
  (testing "Detect type head"
    (let [type (get-type-package exmpl-header)]
      (is (= type :header)))))

(deftest detect-type-test-2
  (testing "Detect type package"
    (let [type (get-type-package exmpl-package)]
              (is (= type :package)))))

(deftest detect-type-test-3
  (testing "Detect type invalid data"
    (let [type (get-type-package exmpl-package-invalid)]
              (is (= type nil)))))

(comment
  (run-all-tests))

;;todo more tests

src/gpsservices/utils/hexlify.clj
(ns gpsservices.utils.hexlify
  (:require [gloss.io :as gio])
  (:use [gloss.data.primitives :refer :all]))

(defprotocol Hexl
  (hexl-hex [val])
  (hexl-char [char]))

(extend-type Number
  Hexl
  (hexl-hex [i]
    (let [rtnval (Integer/toHexString (if (< i 0) (+ 256 i) i)) ]
      (if (< (count rtnval) 2) (str "0" rtnval) rtnval)))
  (hexl-char [b]
    (let [v (if (< b 0) (+ 256 b) b)
          c  (char v)]
      (if  (and (< v 128 )(Character/isLetter c)) (.toString c) "."))))


(extend-type Character
  Hexl
  (hexl-hex [char]
    (hexl-hex (int (.charValue char))))
  (hexl-char [char]
    (hexl-char (int (.charValue char)))))

(defn byte-reverse [^bytes bytes]
  (byte-array (reverse bytes)))

(defn subbytes
  "slice byte-array"
  ([bytes start] (subbytes bytes start (count bytes)))
  ([bytes start end]
   (let [vec-bytes (vec bytes) vec-sub (subvec vec-bytes start end)]
     (byte-array vec-sub))))

(defn sub-and-reverse [^bytes bytes start end]
  (let [sub (subbytes bytes start end)]
    (byte-reverse sub)))

(defn str-to-bytes [str]
  (.getBytes str "UTF-8"))

(defn bytes->ubutes [^bytes bytes]
  (for [val bytes] (gio/decode :ubyte (byte-array [val]))))

(defn ubytes->butes [^bytes bytes]
  (for [val bytes] (gio/decode :byte (byte-array [val]))))

(defn vec-to-bytes [vec]
  "example hex-vec [0xFF 0xFF 0xFF 0xFF 0xFF]"
  (byte-array vec))

(defn concat-bytes [nested-bytes]
  (byte-array (mapcat seq nested-bytes)))

(defn hexify-as-str
  "Convert byte sequence to hex string"
  [coll]
  (let [hex [\0 \1 \2 \3 \4 \5 \6 \7 \8 \9 \a \b \c \d \e \f]]
    (letfn [(hexify-byte [b]
                         (let [v (bit-and b 0xFF)]
                           [(hex (bit-shift-right v 4)) (hex (bit-and v 0x0F))]))]
      (apply str (mapcat hexify-byte coll)))))

(defn get-byte [^bytes bytes i]
  (byte->ubyte (aget bytes i)))

(defn first-byte [^bytes bytes]
  (get-byte bytes 0))

(defn last-byte [^bytes bytes]
  (let [size (dec (count bytes))]
    (get-byte bytes size)))

(defn hexify-str [s]
  (hexify-as-str (.getBytes s)))

(defn- byte-array? [value]
  (= (Class/forName "[B") (class value)))

(defn int->int8 [^Integer num]
  (first (bytes->ubutes (vec (byte-array [num])))))

(defn hexlify
  "Perform similar to hexlify in emacs.  Accept a seq of bytes and
  convert it into a seq of vectors.  The first element of the vector is a
  seq of 16 strings for the HEX value of each byte.  The second element
  of the vector is a seq of the printable representation of the byte and the
  third elevment of thee vector is a seq of the integer value for each
  byte.  Works for chars as well."
  ([bytes] (hexlify bytes 16))
  ([bytes size]
   (let [parts (partition-all size bytes)]
     (for [part parts]
       [ (map hexl-hex part) (map hexl-char part) (map int part) ]))))

(defn hexlify-chars
  "Convert the bytes into a string of printable chars
   with . being used for unprintable chars"
  [^bytes bytes]
  (let [chars (mapcat second (hexlify bytes))]
    (apply str chars)))

src/gpsservices/autolink2.clj
(ns gpsservices.autolink2
  (:gen-class)
  (:require [clojure.core.async :refer [timeout buffer alts!! alts! close! onto-chan go go-loop chan <! >! >!! <!! thread]]
            [manifold.stream :as s]
            [gpsservices.utils.hexlify :as hx]
            [aleph.tcp :as tcp]
            [gloss.core :as g]
            [gloss.io :as gio])
  (:use [gloss.data.primitives :refer :all])
  (:import (java.util UUID)))

(defprotocol ISockEvents
  (on-error [this session err])
  (on-open [this session])
  (on-message [this session msg])
  (on-close [this session]))

(defrecord Session [id chan user-stream info last-packet car-id SockEvents])

(defrecord DefEvents []
  ISockEvents
  (on-error [this session err]
    (println "ISockEvents-error:" (.getMessage err)) this)
  (on-open [this session]
    (println "ISockEvents-open:") this)
  (on-message [this session msg]
    (println "ISockEvents-message:" msg " car-id > " @(:car-id session)) this)
  (on-close [this session]
    (println "ISockEvents-close: car-id >" @(:car-id session)) this))

(g/defcodec utf-frame (g/string :utf-8))

(g/defcodec header-frame (g/ordered-map :car-id :uint64 :protocol :ubyte :header :ubyte))

(defonce server (atom nil))

(defonce users (atom {}))

(defn- gen-uuid [] (str (UUID/randomUUID)))

(defn register-user [session]
  (let [id (:id session)]
    (swap! users assoc id session)))

(defn unregister-user [session]
  (let [id (:id session)
        SockEvents (:SockEvents session)
        user-stream (:user-stream session)]
    (.on-close SockEvents session)
    (s/close! user-stream)
    (swap! users dissoc id)))

(defn full-head? [^bytes data]
  (and (= (count data) 10) (= 0xff (hx/first-byte data)) (= 0x00 (hx/last-byte data))))

(defn full-pack? [^bytes data]
  (and (= 0x5b (hx/first-byte data)) (= 0x5d (hx/last-byte data))))

(defn valid-data? [^bytes data]
  (or (full-head? data) (full-pack? data)))

(defn get-type-package [^bytes data]
  (when data (if (full-head? data) :header
                                   (if (full-pack? data) :package))))

(defn- sub-decode [^bytes bytes start end frame]
  (let [sub (hx/sub-and-reverse bytes start end)]
    (gio/decode frame sub)))

(defn get-frames
  "return all frames from packages data"
  ([^bytes data] (get-frames data 9 []))
  ([^bytes data p-start-offset coll]
   (let [frame-length (sub-decode data 3 5 :uint16)
         p-end-offset (+ p-start-offset frame-length)
         data-count (- (count data) 2)                      ;; удаляем последние два байта так как это CS и end-flag
         frame (hx/subbytes data p-start-offset p-end-offset)
         coll (into coll [frame])]
     (if (< p-end-offset data-count)
       (get-frames data p-end-offset coll)
       coll))))

(defn- bytes->point [^bytes bytes]
  (gio/decode :float32 (hx/byte-reverse bytes)))

(defn- decode-info [^bytes bytes]
  (let [speed (* (hx/get-byte bytes 3) 1.852)
        sat-bin-part (partition 4 (Integer/toString (hx/get-byte bytes 2) 2) )
        sat-GPS-count (Integer/parseInt (apply str (first sat-bin-part)) 2)
        sat-GLONASS-count (Integer/parseInt (apply str (last sat-bin-part)) 2)]
    {:speed speed
     :satellite-count (+ sat-GPS-count sat-GLONASS-count)
     :altitude        (* (hx/get-byte bytes 1) 10)
     :course          (* (hx/get-byte bytes 0) 2)}))


(defn data-sum [^bytes data]
  (let [count-data (count data)]
    (loop [i 0 sum 0]
      (let [val (hx/get-byte data i)
            new-sum (+ sum val)]
        (if (< i (dec count-data))
          (recur (inc i) new-sum) new-sum)))))

(defn valid-CS? [^bytes data]
  (let [count-data (count data)
        origin-CS (aget data (- count-data 2))
        unixtime (hx/subbytes data 5 9)
        frames (get-frames data)
        unixtime-sum (data-sum unixtime)
        frames-sum (reduce + (map data-sum frames))
        CS (hx/int->int8 (+ unixtime-sum frames-sum))]
    (= origin-CS CS)))

(defn- decode-package-frame [^bytes frame]
  (let [frame-parts (partition 5 frame)
        pull (into {} (for [part frame-parts :let [[id & rest] (vec part)]]
                        [id (byte-array rest)]))
        info (decode-info (pull 5))]
    (merge {:point {:lat (bytes->point (pull 3))
                    :lon (bytes->point (pull 4))}
            :status-bytes (pull 9)} info)))

(defn decode-packages [^bytes data]
  (let [pack-num (sub-decode data 1 2 :ubyte)
        unixtime (sub-decode data 5 9 :uint32)
        frames (get-frames data)]
    (when (valid-CS? data)
      {:type :package
       :pack-num pack-num
       :unixtime unixtime
       :data     (vec (map decode-package-frame frames))})))

(defn gen-response
  "generate response for autolink2 device"
  ([] (gen-response 0x00))
  ([pack-num]
   (let [pack-num (or pack-num 0xFE)]
     (byte-array [0x7B 0x00 pack-num 0x7D]))))


(defn decode-data
  "parse binary autolink2 data"
  [^bytes data]
  (let [type (get-type-package data)]
    (case type
      :header (let [data (hx/byte-reverse data)
                    decode-header (dissoc (gio/decode header-frame data)) ]
                    (merge decode-header {:type :header}))
      :package (decode-packages data))))

(defn- pack-reader [^bytes data ^bytes new-data]
  (let [type-new-data (get-type-package new-data)
        type-last-data (get-type-package data)
        concat-data (hx/concat-bytes [data new-data])
        size-last-data (count data)]
    (if type-new-data new-data
                      (if (> size-last-data 1500)
                        (do (throw (Exception. "exceeded the minimum (1500b) buf size")) nil)
                        (if type-last-data new-data concat-data)))))

(defn- channel-handler [session]
  (let [{chan        :chan
         user-stream :user-stream
         last-packet :last-packet
         car-id      :car-id
         SockEvents  :SockEvents} session]
    (go-loop []
      (if-let [ch-data (<! chan)]
        (let [res (try
                    (swap! last-packet pack-reader ch-data)
                    (let [last-packet @last-packet]
                      (when (valid-data? last-packet)
                        (let [decrypted-data (decode-data last-packet) type (get-type-package last-packet)]
                          (when-let [new-car-id (:car-id decrypted-data)]
                            (reset! car-id new-car-id))
                          (case type
                            :header (s/put! user-stream (gen-response))
                            :package (if @car-id
                                       (s/put! user-stream (gen-response (:pack-num decrypted-data)))
                                       (throw (Exception. "header pack is missing"))))
                          (.on-message SockEvents session decrypted-data))))
                    :ok
                    (catch Exception e (.on-error SockEvents session e)
                                       (unregister-user session)
                                       (close! chan)
                                       :error))]
          (when-not (= res :error)
            (recur)))
        (unregister-user session)))) session)

(defn- channel-new
  ([user-stream SockEvents]
   (channel-new user-stream SockEvents {}))
  ([user-stream SockEvents info]
   (channel-new user-stream SockEvents info channel-handler))
  ([user-stream SockEvents info handler]
   (let [id (gen-uuid) chan (chan) session (->Session id chan user-stream info (atom nil) (atom nil) SockEvents)]
     (s/connect user-stream chan)
     (handler session))))

(defn start-server
  ([] (start-server (->DefEvents)))
  ([SockEvents & [options]]
   (let [port (:port options 7779)
         srv (tcp/start-server (fn [user-stream info]
                                 (let [session (channel-new user-stream SockEvents info)]
                                   (register-user session)
                                   (.on-open SockEvents session)))
                               {:port port})]
     (printf "gpsservices.autolink2 server started. listen on %s\n" port)
     (reset! server srv))))

(defn stop-server []
  "close all sessions and stop server"
  (doseq [[_ user] @users]
    (close! (:chan user)))
  (when-let [srv @server]
    (.close srv)))
